src/linear_algebra/trace.lean

```diff
@@ -115,6 +116,37 @@ end
 section
 variables (R : Type u) [field R] {M : Type v} [add_comm_group M] [module R M]
 
+/-- The trace of a linear map correspond to the contraction pairing under the isomorphism
+ `End(M) ≃ M* ⊗ M`-/
+@[simp] lemma trace_eq_contract [finite_dimensional R M] :
+  (linear_map.trace R M) ∘ₗ (dual_tensor_hom R M M) = contract_left R M :=
+begin
+  have b := fin_basis R M,
+  apply basis.ext (basis.tensor_product (b.dual_basis) b),
+  rintros ⟨i, j⟩,
+  simp only [dual_tensor_hom_equiv_apply, basis.coord_apply, linear_equiv.coe_coe,
+    basis.repr_self, function.comp_app, contract_left_apply, basis.tensor_product_apply,
+    basis.coe_dual_basis, coe_comp],
+  rw [trace_eq_matrix_trace R b, dual_tensor_hom_basis],
+  by_cases hij : i = j,
+  { rw [hij, matrix.std_basis_matrix.trace_same], simp },
+  rw matrix.std_basis_matrix.trace_zero j i (1:R) hij,
+  simp [finsupp.single_eq_pi_single, hij],
+end
+
+/-- The trace of a linear map correspond to the contraction pairing under the isomorphism
+ `End(M) ≃ M* ⊗ M`-/
+lemma trace_eq_contract' [finite_dimensional R M] :
+  (linear_map.trace R M) = (contract_left R M) ∘ₗ ↑(dual_tensor_hom_equiv R M M).symm :=
+begin
+  rw [←trace_eq_contract, comp_assoc],
+  rw [←coe_dual_tensor_hom_equiv_to_linear],
+  suffices H : ↑(dual_tensor_hom_equiv R M M) ∘ₗ ↑(dual_tensor_hom_equiv R M M).symm = id,
+  { rw [H], simp, },
+  apply linear_map.coe_injective,
+  simp [-coe_dual_tensor_hom_equiv_to_linear],
+end
```

Comment:
This proof should be a 1-liner. Please add a lemma `semilinear_map.eq_comp_symm` which says that
```
f = g ∘ₗ e.symm ↔ f ∘ e = g
```
(and similarly `eq_symm_comp`, `comp_symm_eq` and `symm_comp_eq`).
With that lemma in place, the proof should be
```
by rw [eq_comp_symm, coe_dual_tensor_hom_equiv, trace_eq_contract]
```

Reply:
I haven't looked at this PR in detail, but I think the lemma should be called `linear_map.eq_comp_symm` -- we generalized linear maps to semilinear maps, but it's still called `linear_map`!